### PR TITLE
[1.0] Harden ordered_diff

### PR DESCRIPTION
**Pull request by heifner**
`ordered_diff` is used to process the finalizer and proposer policy diffs from block header extension `finality_extension`. Since the `finality_extension`, is provided to a node to validate, validate the diff during `apply_diff`.

Related to #691

### CODE REVIEW

**Review by greg7mdp — APPROVED** [review]

File: libraries/libfc/include/fc/container/ordered_diff.hpp [hunk 1]
```diff
@@ -36,50 +41,50 @@ class ordered_diff {
...
             if (source[s] == target[t]) {
                // nothing to do, skip over
-               assert(s <= std::numeric_limits<SizeType>::max());
-               assert(t <= std::numeric_limits<SizeType>::max());
                ++s;
                ++t;
             } else { // not equal
                if (s == source.size() - 1 && t == target.size() - 1) {
+                  FC_ASSERT(result.remove_indexes.size() < MAX_NUM_ARRAY_ELEMENTS);
+                  FC_ASSERT(result.insert_indexes.size() < MAX_NUM_ARRAY_ELEMENTS);
```
Comment (greg7mdp) [inline libraries/libfc/include/fc/container/ordered_diff.hpp:54]: I'd be tempted to also move these two right after the `while` so they are not repeated throughout.

**Review by greg7mdp — APPROVED** [review]

File: libraries/libfc/include/fc/container/ordered_diff.hpp [hunk 2]
```diff
@@ -36,50 +41,44 @@ class ordered_diff {
...
                   result.insert_indexes.emplace_back(t, target[t]);
                   ++s;
                   ++t;
                } else if (t + 1 < target.size() && source[s] == target[t + 1]) {
                   // source equals next target, insert current target
-                  assert(t <= std::numeric_limits<SizeType>::max());
                   result.insert_indexes.emplace_back(t, target[t]);
                   ++t;
                } else { // source[s + 1] == target[t]
                   // target matches next source, remove current source
```
Comment (arhag) [inline libraries/libfc/include/fc/container/ordered_diff.hpp:71]: Comments here need to be updated.

File: libraries/libfc/include/fc/container/ordered_diff.hpp [hunk 3]
```diff
@@ -94,15 +93,23 @@ class ordered_diff {
...
+                   ("idx", index)("o", offset)("s", container.size()));
          container.erase(container.begin() + index + offset);
          --offset;
       }
 
       // Insert into the source based on diff.insert_indexes
       for (auto& [index, value] : diff.insert_indexes) {
+         FC_ASSERT(index <= container.size(), "diff.insert_indexes index ${idx} not in range ${s}",
+                   ("idx", index)("s", container.size()));
+         FC_ASSERT(container.size() < MAX_NUM_ARRAY_ELEMENTS);
```
Comment (arhag) [inline libraries/libfc/include/fc/container/ordered_diff.hpp:112]: This `FC_ASSERT(container.size() < MAX_NUM_ARRAY_ELEMENTS)` can be removed.

File: libraries/libfc/test/test_ordered_diff.cpp [hunk 4]
```diff
@@ -103,7 +103,7 @@ BOOST_AUTO_TEST_CASE(ordered_diff_test) try {
       BOOST_TEST(source == target);
    }
    { // full
-      vector<uint8_t> source(std::numeric_limits<uint8_t>::max()+1);
+      vector<uint8_t> source(std::numeric_limits<uint8_t>::max()-1);
```
Comment (arhag) [inline libraries/libfc/test/test_ordered_diff.cpp:106]: This should be reverted.

**Review by arhag — CHANGES_REQUESTED** [review]
Also clean up `static_assert`s that check the max representable value of `SizeType` is greater than or equal to `config::max_finalizers - 1` or `config::max_proposers - 1` .

**Review by arhag — CHANGES_REQUESTED** [review]

**Review by arhag — APPROVED** [review]

File: libraries/libfc/include/fc/container/ordered_diff.hpp [hunk 5]
```diff
@@ -52,6 +63,7 @@ class ordered_diff {
                   result.insert_indexes.emplace_back(t, target[t]);
                   ++s;
                   ++t;
+                  assert(s > 0 && t > 0);
```
Comment (greg7mdp) [inline libraries/libfc/include/fc/container/ordered_diff.hpp:66]: considering that `s` and `t `are of type `size_t`, I don't think these asserts are necessary!

**Review by greg7mdp — CHANGES_REQUESTED** [review]

File: libraries/libfc/include/fc/container/ordered_diff.hpp [hunk 6]
```diff
@@ -60,28 +72,34 @@ class ordered_diff {
...
                   result.insert_indexes.emplace_back(t, target[t]);
                   ++t;
-               } else { // source[s + 1] == target[t]
-                  // target matches next source, remove current source
-                  assert(s <= std::numeric_limits<SizeType>::max());
+                  assert(t > 0);
+               } else {
+                  // not misalignment by one and source not equal to next target, so remove from source
+                  // may be inserted later by other conditions if needed
+                  assert(t <= std::numeric_limits<SizeType>::max());
```
Comment (greg7mdp) [inline libraries/libfc/include/fc/container/ordered_diff.hpp:85]: why `t`?

**Review by greg7mdp — APPROVED** [review]

File: libraries/libfc/include/fc/container/ordered_diff.hpp (hunk 5 above)
Comment (heifner) [reply libraries/libfc/include/fc/container/ordered_diff.hpp:66]: Makes sense, I'll remove them.

File: libraries/libfc/include/fc/container/ordered_diff.hpp (hunk 6 above)
Comment (heifner) [reply libraries/libfc/include/fc/container/ordered_diff.hpp:85]: Thanks, `s` obviously

Comment (arhag) [reply libraries/libfc/include/fc/container/ordered_diff.hpp:85]: Nice catch

**Review by arhag — APPROVED** [review]

**Review by greg7mdp — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: STABILITY
category: INTERNALS
summary: Additional checks to harden the processing of the finalizer and proposer policy diffs from block header. extension.
Note:end

**Comment by ericpassmore** [comment]
Note:start
category: Other
component: Internal
summary: Additional checks to harden the processing of the finalizer and proposer policy diffs from block header extension.
Note:end

**Pull request merged**